---
title: 'Supplementary Figure 7 d,e: DDLS deconvolution of TCGA BRCA data, correlation with cDC1 and survival prediction of CD4 Trms'
author: "Ignacio Heras"
date: "`r paste0(Sys.Date())`"
documentclass: article
output:
  html_document: 
    number_sections: yes
    self_contained: no
    theme: united
    toc: yes
    toc_depth: 5  
    toc_float: true
    collapsed: no
  pdf_document:
    fig_caption: yes
    keep_tex: yes
    number_sections: yes
    toc: yes
    toc_depth: 4
urlcolor: blue
classoption: a4paper
---

```{r setup, include=FALSE}
# load packages
library(limma)
library(stringr)
library(ggpubr)
library(RColorBrewer)
library(genefilter)
library(gage)
library(fgsea)
library(MCPcounter)
library(tidyr)
library(dplyr)
library(readxl)
library(purrr)
suppressMessages(library("Seurat"))
suppressMessages(library("purrr"))
suppressMessages(library("ggplot2"))
suppressMessages(library("stringr"))
suppressMessages(library("readr"))
suppressMessages(library("digitalDLSorteR"))
suppressMessages(library("SingleCellExperiment"))
suppressMessages(library("SummarizedExperiment"))
suppressMessages(library("corrplot"))
suppressMessages(library("ggpubr"))
suppressMessages(library("survival"))
suppressMessages(library("ggsurvfit"))
suppressMessages(library("gtsummary"))

# Setup of analysis
projectPath <- here::here()
dataPath <- here::here("data")
dataSKCMPath <- here::here("data/skcm_tcga/gdac.broadinstitute.org_SKCM.mRNAseq_Preprocess.Level_3.2016012800.0.0")
dataBCPath <- here::here("data/brca_tcga/gdac.broadinstitute.org_BRCA.mRNAseq_Preprocess.Level_3.2016012800.0.0")

metadataPath <- here::here("metadata")
objectsPath <- here::here("objects")
reportsPath <- here::here("reports")
plotsPath <- here::here("plots")
prefix <- "trial_2"

## Settings to knitr
knitr::opts_chunk$set(
    fig.align = "center",
    fig.path = plotsPath,
    fig.pos = "H",
    message = FALSE,
    warning = FALSE,
    comment = NA,
    echo = FALSE,
    dev = c("jpeg"),
    dpi = 300
)

```


# Loading single-cell data

## Load seurat object

```{r}
seurat.all <- readRDS(file.path(objectsPath, "trial5.seurat.final.all.rds"))
```

## Create SingleCellExperiment object

```{r}
sce.all <- SingleCellExperiment(
  assays = list(counts = seurat.all@assays$RNA@counts),
  colData = DataFrame(
    Cell_ID = colnames(seurat.all),
    row.names = colnames(seurat.all),
    Final_Cell_Type_01 = seurat.all$Final_Cell_Type_01,
    Final_Cell_Type_02 = seurat.all$Final_Cell_Type_02,
    Final_Cell_Type_03 = seurat.all$Final_Cell_Type_03
  )
)
rowData(sce.all) <- DataFrame(
  SYMBOL = rownames(sce.all),
  row.names = rownames(sce.all)
)
```

# Loading bulk data

```{r}
BC.breast <- file.path(dataBCPath, "BRCA.mRNAseq_raw_counts.txt") %>% 
  data.table::fread(sep = "\t", header = TRUE) %>% as.data.frame()

genesMetadata.BC <- data.frame(
  SYMBOL = BC.breast[[1]] %>% str_split_i(pattern = "\\|", i = 1),
  NCBI = BC.breast[[1]] %>% str_split_i(pattern = "\\|", i = 2)
)
BC.breast[[1]] <- NULL
BC.breast <- BC.breast %>% as.matrix()
BC.breast.agg <- aggregate(BC.breast, by = list(genesMetadata.BC$SYMBOL), FUN = sum)
rownames(BC.breast.agg) <- BC.breast.agg$Group.1
BC.breast.agg[[1]] <- NULL
BC.breast.agg <- as.matrix(BC.breast.agg)
BC.breast.agg <- BC.breast.agg[-1, ]
```

```{r}
TCGA.breast.sce <- SummarizedExperiment(
  BC.breast.agg, 
  colData = DataFrame(
    Sample.ID = colnames(BC.breast.agg),
    row.names = colnames(BC.breast.agg)
  ),
  rowData = DataFrame(
    SYMBOL = rownames(BC.breast.agg),
    row.names = rownames(BC.breast.agg)
  )
)
```


# Creating digitalDLSorter object and deconvolution workflow

Check <https://diegommcc.github.io/digitalDLSorteR/articles/realModelWorkflow-hq.html>.

```{r}
DDLS.breast <- createDDLSobject(
  sc.data = sce.all,
  sc.cell.ID.column = "Cell_ID",
  sc.gene.ID.column = "SYMBOL",
  sc.cell.type.column = "Final_Cell_Type_03",
  bulk.data = TCGA.breast.sce,
  bulk.sample.ID.column = "Sample.ID",
  bulk.gene.ID.column <- "SYMBOL",
  filter.mt.genes = "^MT-",
  sc.min.mean.counts = 2,
  sc.min.counts = 2,
  sc.min.cells = 10,
  verbose = TRUE, 
  project = "Breast-Cancer-Cell-02"  
)
```

```{r}
prop.design <- sce.all@colData %>% as.data.frame() %>% 
  mutate(Patient = str_split_i(rownames(.), "_", 1)) %>% 
  group_by(Patient, Final_Cell_Type_03) %>% summarize(Total = n()) %>% 
  mutate(Prop = (Total / sum(Total)) * 100) %>% group_by(Final_Cell_Type_03) %>% 
  summarise(Prop_Mean = ceiling(mean(Prop)), Prop_SD = ceiling(sd(Prop))) %>% 
  mutate(
    from = Prop_Mean, 
    to.1 = Prop_Mean * (Prop_SD * 2),
    to = ifelse(to.1 > 100, 100, to.1),
    to = ifelse(to < 10, 30, to),
    to.1 = NULL, Prop_Mean = NULL, Prop_SD = NULL
  )
```


```{r}
set.seed(123)
DDLS.breast <- generateBulkCellMatrix(
  object = DDLS.breast,
  cell.ID.column = "Cell_ID",
  cell.type.column = "Final_Cell_Type_03",
  prob.design = prop.design,
  num.bulk.samples = 40000, 
  prob.sparsity = 1,
  min.zero.prop = 16,
  n.cells = 50,
  verbose = TRUE
  
) %>% simBulkProfiles(threads = 4)
```

```{r}
DDLS.breast <- trainDDLSModel(
  DDLS.breast, verbose = TRUE, 
  num.epochs = 100, scaling = "rescale", 
  norm.batch.layers = FALSE, 
  activation.fun = "sigmoid"
)
```

```{r}
DDLS.breast <- calculateEvalMetrics(object = DDLS.breast)
```

```{r, fig.height=10, fig.width=12, eval = FALSE}
corrExpPredPlot(
  DDLS.breast,
  color.by = "CellType",
  facet.by = "CellType",
  corr = "both", 
  size.point = 0.5
)
```

# Loading new bulk data: version with more samples

```{r}
raw.matrix.2 <- file.path(
  dataPath, "brca_tcga", 
  "gdac.broadinstitute.org_BRCA.mRNAseq_Preprocess.Level_3.2016012800.0.0",
  "BRCA.uncv2.mRNAseq_raw_counts.txt"
) %>% read_tsv()
genesMetadata.BC.raw.2 <- data.frame(raw = raw.matrix.2[[1]][-1]) %>% 
  mutate(
    SYMBOL = str_split_i(raw, pattern = "\\|", i = 1), 
    NCBI = str_split_i(raw, pattern = "\\|", i = 2)
  )
raw.matrix.2[[1]] <- NULL
raw.matrix.2 <- raw.matrix.2[-1, ]
raw.matrix.2.num <- sapply(raw.matrix.2, \(x) as.numeric(x))

raw.matrix.2.num.agg <- aggregate(
  raw.matrix.2.num, by = list(genesMetadata.BC.raw.2$SYMBOL), FUN = sum
)
genesMetadata.BC.raw.2.agg <- data.frame(SYMBOL = raw.matrix.2.num.agg[[1]][-1])
raw.matrix.2.num.agg[[1]] <- NULL
raw.matrix.2.num.agg <- raw.matrix.2.num.agg[-1, ]
rownames(raw.matrix.2.num.agg) <- genesMetadata.BC.raw.2.agg[["SYMBOL"]]
```

```{r}
TCGA.breast.sce.raw.2 <- SummarizedExperiment(
  raw.matrix.2.num.agg, 
  colData = DataFrame(
    Sample.ID = colnames(raw.matrix.2.num.agg),
    row.names = colnames(raw.matrix.2.num.agg)
  ),
  rowData = DataFrame(
    SYMBOL = rownames(raw.matrix.2.num.agg),
    row.names = rownames(raw.matrix.2.num.agg)
  )
)
DDLS.breast <- DDLS.breast %>% loadDeconvData(
  data = TCGA.breast.sce.raw.2, name.data = "Raw-2"
)
deconv.counts <- as.matrix(assay(deconv.data(DDLS.breast, "Raw-2")))
deconv.counts <- deconv.counts[DDLS.breast@trained.model@features, ]
deconv.counts <- log2(digitalDLSorteR:::.cpmCalculate(x = deconv.counts + 1))
deconv.data <- digitalDLSorteR:::rescale.function(t(deconv.counts))

results <- model(DDLS.breast@trained.model) %>% predict(
  deconv.data,
  verbose = TRUE
)  
rownames(results) <- rownames(deconv.data)
colnames(results) <- cell.types(DDLS.breast@trained.model)
```

```{r}
saveRDS(results, file.path(objectsPath, "deconv.results.new.samples-1200.old.model.rds"))
```



#Load digital sorter results

# digitalDLSorteR results

```{r}
deconv.results <- readRDS(file.path(objectsPath, "deconv.results.new.samples-1200.old.model.rds"))
```

#Check correlations

```{r}

deconv.data.scatter <- deconv.results %>% as.data.frame() %>% 
  filter(!grepl("1[10]$", rownames(.)))
colnames(deconv.data.scatter) <- str_replace_all(colnames(deconv.data.scatter), c(" " = "_", "\\+" = ""))
deconv.data.scatter <- deconv.data.scatter %>% as.data.frame() %>% mutate(
    SAMPLE_ID = rownames(.)
  ) %>% mutate(SAMPLE_ID = str_sub(SAMPLE_ID, 1, nchar(SAMPLE_ID) - 3))


p1 <- deconv.data.scatter %>% ggscatter(
    x = "cDC1", y = "CD4_Trm",
    color = "black", 
    size = 1,
    add = "reg.line", conf.int = TRUE,
    cor.coef = TRUE, cor.method = "pearson"
) 

annotate_figure(
  p1, top = text_grob("BRCA - correlation to CD4 Trm Score digital sorter", face = "bold", size = 14)
)
```

## Combine with filtered clinical data


```{r}
clinicalMetadata.BC.f.deconv <- deconv.data.scatter %>% left_join(
  clinicalMetadata.BC.f,
  by = c("SAMPLE_ID" = "bcr_patient_barcode")
) %>% filter(!is.na(CAFs))
```

#Create quartiles and tertiles
```{r}
clinicalMetadata.BC.f.deconv <- clinicalMetadata.BC.f.deconv %>% mutate(
    CD4_Trm_tertiles = cut(
     CD4_Trm, 
     breaks = quantile(
       CD4_Trm, probs = seq(0, 1, by = 1/3)
     ), 
     include.lowest = TRUE, labels = c("Low", "Medium", "High")
   )
) 


map(
  colnames(clinicalMetadata.BC.f.deconv) %>% grep("CD4_Trm_tertiles", ., value = T), 
  \(cond) {
    survfit2(
      as.formula(paste0("Surv(OS.time, OS) ~ ", cond)), 
      data = clinicalMetadata.BC.f.deconv
    ) %>% ggsurvfit() + add_risktable() + 
      ggtitle(paste0("Survival curves split by median cell abundance [", cond, "]"))
  }
)
```

# Statistical analysis

```{r, fig.height=8, fig.width=10}

res.surv.all <- survdiff(
  as.formula(paste0("Surv(OS.time, OS) ~ CD4_Trm_tertiles")), 
  data = clinicalMetadata.BC.f.deconv %>% mutate(
    CD4_Trm_tertiles = factor(CD4_Trm_tertiles, levels = c("Low",  "High"))
  )
) 
res.surv.all
```


